server/src/main/java/org/elasticsearch/index/engine/SoftDeletesPolicy.java

```diff
@@ -0,0 +1,113 @@
+ * under the License.
+ */
+
+package org.elasticsearch.index.engine;
+
+import org.apache.lucene.document.LongPoint;
+import org.apache.lucene.search.Query;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.index.mapper.SeqNoFieldMapper;
+import org.elasticsearch.index.seqno.SequenceNumbers;
+import org.elasticsearch.index.translog.Translog;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.LongSupplier;
+
+/**
+ * A policy that controls how many soft-deleted documents should be retained for peer-recovery and querying changes purpose.
+ */
+final class SoftDeletesPolicy {
+    private final LongSupplier globalCheckpointSupplier;
+    private int retentionLockCount;
+    private long checkpointOfSafeCommit;
+    private long minRequiredSeqNoForRecovery;
+    private long retentionOperations;
+
+    SoftDeletesPolicy(LongSupplier globalCheckpointSupplier, long retentionOperations) {
+        this.globalCheckpointSupplier = globalCheckpointSupplier;
+        this.retentionOperations = retentionOperations;
+        this.checkpointOfSafeCommit = SequenceNumbers.NO_OPS_PERFORMED;
+        this.minRequiredSeqNoForRecovery = checkpointOfSafeCommit;
+        this.retentionLockCount = 0;
+    }
+
+    /**
+     * Updates the number of soft-deleted prior to the global checkpoint to be retained
+     * See {@link org.elasticsearch.index.IndexSettings#INDEX_SOFT_DELETES_RETENTION_OPERATIONS_SETTING}
+     */
+    synchronized void setRetentionOperations(long retentionOperations) {
+        this.retentionOperations = retentionOperations;
+    }
+
+    /**
+     * Sets the local checkpoint of the current safe commit.
+     * All operations whose seqno are greater than this checkpoint will be retained until the new checkpoint is advanced.
+     */
+    synchronized void setCheckpointOfSafeCommit(long newCheckpoint) {
+        if (newCheckpoint < this.checkpointOfSafeCommit) {
+            throw new IllegalArgumentException("Local checkpoint can't go backwards; " +
+                "new checkpoint [" + newCheckpoint + "]," + "current checkpoint [" + checkpointOfSafeCommit + "]");
+        }
+        this.checkpointOfSafeCommit = newCheckpoint;
+        updateMinRequiredSeqNoForRecovery();
+    }
+
+    private void updateMinRequiredSeqNoForRecovery() {
+        assert Thread.holdsLock(this) : Thread.currentThread().getName();
+        if (retentionLockCount == 0) {
+            this.minRequiredSeqNoForRecovery = checkpointOfSafeCommit;
+        }
+    }
+
+    /**
+     * Acquires a lock on soft-deleted documents to prevent them from cleaning up in merge processes. This is necessary to
+     * make sure that all operations after the local checkpoint of the safe commit are retained until the lock is released.
+     * This is a analogy to the translog's retention lock; see {@link Translog#acquireRetentionLock()}
+     */
+    synchronized Releasable acquireRetentionLock() {
```

Comment:
this is not used and not tested - I understand it will be used in the future for the recoveries, but can you please add them along with the usage? then we can see if it works.

Reply:
it was used in tests and I asked the same question offline ;)